File: Java/src/HowToSolveCodingProblems/InterviewGoogle.java
package HowToSolveCodingProblems;

//find two numbers whose sum matches the target as parameter
//input Array with nunmbers -- target int sum
//output two numbers from array true / false


import java.util.HashSet;
import java.util.Set;

public class InterviewGoogle {

    public static void main(String[] args) {



        int[] array1 = {1,2,3,9}; // 8
        int[] array2 = {6,4,3,2,1,7}; // 9



        System.out.println(hasPairWithSum3(array2, 9));


    }

    public static boolean hasPairWithSum(int[] array, int target){

        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j <array.length ; j++) {
                if (array[i] + array[j] == target){
                    System.out.println(target);
                    System.out.println(array[i] + " : " + array[j]);
                    return true;
                }
            }
        }
        return false;
    }

    public static boolean hasPairWithSum2(int[] array, int target){
        int anterior = 0;
        for (int i = 0; i < array.length; i++) {
            int inicial = array[i];
            anterior +=inicial;
            int second = array[array.length-1];

            if ( (inicial + second) == target || ((anterior + inicial) == target) && anterior != target){
                System.out.println(inicial + " : " + second);
                System.out.println(second + " : " + anterior);
                return true;
            }

        }
        return false;
    }

    public static boolean hasPairWithSum3(int[] array, int target){

        Set<Integer> mySet = new HashSet<>();
        for (int i = 0; i < array.length ; i++) {
            if (mySet.contains(i)) {
                return true;
            }
            mySet.add(target - array[i]);
        }
        return false;
    }
}

File: JavaScript/HowToSolveCodingProblems/interviewGoogle.js
//find two numbers whose sum matches the target as parameter
//input Array with nunmbers -- target int sum
//output two numbers from array true / false

const arrayNumbers = [6,4,8,5,1,3,7]
const target = 7;

// brute approach
const hasPairWithSum = (numbers = [], target) =>{

    for(let i = 0; i < numbers.length; i++){
        for(let j = 1; j < numbers.length -1; j++){
            
            if(numbers[i] + numbers[j] == target ){
                console.log('puto toto mkon ya lo encontre')
                return true;
            }
        }
    }

    return false;

}

// O(n^2)
console.log(hasPairWithSum(arrayNumbers, target));


// approach with map
const hasPairWithSum2 = (numbers = [], target) =>{

    const myMap = new Map();

    numbers.forEach( elem => myMap.set(elem, target));

    for (let i = 0; i < numbers.length; i++){
        const rest = (myMap.get(numbers[i]) - numbers[i]);
        if(myMap.get(rest) ){
            console.log(myMap.get(rest) - rest, ' + ', rest, '=', target)
            return true;
        }
        
    }
    
    return false;

}

// O(n)
// O(n) Space Complexity
console.log(hasPairWithSum2(arrayNumbers, target));


// Andrei Approach
const hasPairWithSum3 = (numbers = [], target) =>{

    const mySet = new Set();

    const len = numbers.length;
    for (let i = 0; i < len; i++){
        console.log(mySet.has(numbers[i]))
        if(mySet.has(numbers[i])) return true;
        mySet.add(target - numbers[i]);        
    }
    return false;

}

// O(n)
// O(n) Space Complexity
console.log(hasPairWithSum3(arrayNumbers, target));
